File: leetcode/java/0647. Palindromic Substrings.java
package leetcode.java;

// Time complexity: O(N^2)
// Space complexity: O(1)

class Solution {
    public int countSubstrings(String s) {
        int numPalindromes = 0;
        
        for (int i=0; i<s.length(); i++) {

            int left = i;
            int right = i;

            while (left >=0 && right<s.length() && s.charAt(left) == s.charAt(right)) {
                numPalindromes += 1;
                left -= 1;
                right += 1;
            }

            left = i;
            right = i+1;

            while (left >=0 && right<s.length() && s.charAt(left) == s.charAt(right)) {
                numPalindromes += 1;
                left -= 1;
                right += 1;
            }
        }
        return numPalindromes;
    }
}

File: leetcode/python/0647. Palindromic Substrings.py
# Time complexity: O(N^2)
# Space complexity: O(1)

class Solution(object):
    def countSubstrings(self, s):
        res = 0

        for i in range(len(s)):
            
            left = i
            right = i
            while left>=0 and right<len(s) and s[left] == s[right]:
                left -= 1
                right += 1
                res += 1
            
            left = i
            right = i+1
            while left>=0 and right<len(s) and s[left] == s[right]:
                left -= 1
                right += 1
                res += 1

        return res        